File: extensions/vscode/src/nativeSdkBackend.ts
import * as vscode from "vscode";
import * as os from "os";
import * as path from "path";
import * as fs from "fs";
import type { SdkBackend, SdkEntry } from "./sdkProvider";

const FLUTTER_GIT_URL = "https://github.com/flutter/flutter.git";
const SDK_VERSIONS_REL = "flutter_compile/versions";
const RC_FILE = ".flutter_compilerc";
const GLOBAL_SDK_KEY = "global_sdk_version";
const FLUTTER_VERSION_FILE = ".flutter-version";

const SDK_PATH_BLOCK_START =
  "# >>> Added by flutter_compile SDK manager >>>";
const SDK_PATH_BLOCK_END =
  "# <<< Added by flutter_compile SDK manager <<<";

const ENV_FILE = ".flutter_compile_env";
const DEFAULT_SDK_LINK = "default";

/** Returns the user's home directory. */
function homeDir(): string {
  return os.homedir();
}

/** Returns the SDK versions directory (`~/flutter_compile/versions`). */
function versionsDir(): string {
  return path.join(homeDir(), SDK_VERSIONS_REL);
}

/** Returns the absolute install path for a given SDK version. */
function sdkVersionPath(version: string): string {
  return path.join(versionsDir(), version);
}

/** Returns the RC config file path (`~/.flutter_compilerc`). */
function rcFilePath(): string {
  return path.join(homeDir(), RC_FILE);
}

// ---------------------------------------------------------------------------
// RC config helpers
// ---------------------------------------------------------------------------

/** Read the rc config file as a key-value map. */
function readRcConfig(): Map<string, string> {
  const map = new Map<string, string>();
  const filePath = rcFilePath();
  if (!fs.existsSync(filePath)) {
    return map;
  }
  const content = fs.readFileSync(filePath, "utf-8");
  for (const line of content.split("\n")) {
    const colonIndex = line.indexOf(":");
    if (colonIndex !== -1) {
      map.set(line.substring(0, colonIndex), line.substring(colonIndex + 1));
    }
  }
  return map;
}

/** Write a key-value pair to the rc config, preserving existing entries. */
function writeRcConfigKey(key: string, value: string): void {
  const map = readRcConfig();
  map.set(key, value);
  const lines: string[] = [];
  map.forEach((v, k) => lines.push(`${k}:${v}`));
  const dir = path.dirname(rcFilePath());
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(rcFilePath(), lines.join("\n") + "\n", "utf-8");
}

/** Remove a key from the rc config. */
function removeRcConfigKey(key: string): void {
  const filePath = rcFilePath();
  if (!fs.existsSync(filePath)) {
    return;
  }
  const content = fs.readFileSync(filePath, "utf-8");
  const filtered = content
    .split("\n")
    .filter((line) => {
      const colonIndex = line.indexOf(":");
      if (colonIndex === -1) {
        return line.trim().length > 0;
      }
      return line.substring(0, colonIndex) !== key;
    });
  fs.writeFileSync(filePath, filtered.join("\n") + "\n", "utf-8");
}

/** Read a single key value from the rc config. */
function readRcConfigValue(key: string): string | undefined {
  const map = readRcConfig();
  return map.get(key);
}

// ---------------------------------------------------------------------------
// Shell config helpers (PATH management)
// ---------------------------------------------------------------------------

/** Detect the user's shell config file path. */
function shellConfigPath(): string {
  const home = homeDir();
  if (process.platform === "win32") {
    const docs = process.env.USERPROFILE ?? home;
    return path.join(
      docs,
      "Documents",
      "WindowsPowerShell",
      "Microsoft.PowerShell_profile.ps1",
    );
  }
  const shell = process.env.SHELL ?? "";
  const rc = shell.includes("bash")
    ? ".bashrc"
    : shell.includes("zsh")
      ? ".zshrc"
      : ".profile";
  return path.join(home, rc);
}

/** Returns the path to the dedicated env file (~/.flutter_compile_env). */
function envFilePath(): string {
  return path.join(homeDir(), ENV_FILE);
}

/** Source line added to the user's shell RC. */
const SOURCE_LINE = `\n[ -f ~/.${ENV_FILE} ] && source ~/.${ENV_FILE}\n`;
const SOURCE_LINE_WINDOWS = `\nif (Test-Path "$HOME\\.${ENV_FILE}") { . "$HOME\\.${ENV_FILE}" }\n`;

/** Regex to detect the source line already present in shell RC. */
const SOURCE_LINE_RE =
  process.platform === "win32"
    ? /\n?if \(Test-Path "\$HOME\\\.flutter_compile_env"\) \{ \. "\$HOME\\\.flutter_compile_env" \}\n?/
    : /\n?\[ *-f *~\/\.flutter_compile_env *\] *&& *source *~\/\.flutter_compile_env\n?/;

/** Regex that matches the SDK manager PATH block. */
const SDK_PATH_BLOCK_RE = new RegExp(
  "\\n?# >>> Added by flutter_compile SDK manager >>>" +
    "[\\s\\S]*?" +
    "# <<< Added by flutter_compile SDK manager <<<\\n?",
);

/**
 * Idempotently ensure the shell RC has a `source` line for the env file.
 * Also strips any legacy SDK manager PATH blocks from the shell RC as migration.
 */
function ensureSourceLine(): void {
  const configFile = shellConfigPath();
  let contents = "";
  try {
    contents = fs.readFileSync(configFile, "utf-8");
  } catch {
    // File doesn't exist yet
  }

  let changed = false;

  // Migration: remove old SDK manager blocks from shell RC
  if (SDK_PATH_BLOCK_RE.test(contents)) {
    contents = contents.replace(SDK_PATH_BLOCK_RE, "");
    changed = true;
  }

  // Add source line if not already present
  if (!SOURCE_LINE_RE.test(contents)) {
    const line =
      process.platform === "win32" ? SOURCE_LINE_WINDOWS : SOURCE_LINE;
    contents += line;
    changed = true;
  }

  if (changed) {
    const dir = path.dirname(configFile);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    fs.writeFileSync(configFile, contents, "utf-8");
  }
}

/** Build the SDK PATH export block for the given SDK path. */
function buildSdkPathBlock(sdkDir: string, pubCachePath: string): string {
  if (process.platform === "win32") {
    return [
      "",
      SDK_PATH_BLOCK_START,
      `if (-not $env:FLUTTER_COMPILE_SDK) {`,
      `  $env:PATH = "${sdkDir}\\bin;$env:PATH"`,
      `  $env:PATH = "${sdkDir}\\bin\\cache\\dart-sdk\\bin;$env:PATH"`,
      `  $env:PUB_CACHE = "${pubCachePath}"`,
      `} else {`,
      `  $env:PATH = "$env:FLUTTER_COMPILE_SDK\\bin;$env:FLUTTER_COMPILE_SDK\\bin\\cache\\dart-sdk\\bin;$env:PATH"`,
      `  $env:PUB_CACHE = "$env:FLUTTER_COMPILE_SDK\\.pub-cache"`,
      `}`,
      SDK_PATH_BLOCK_END,
      "",
    ].join("\n");
  }
  return [
    "",
    SDK_PATH_BLOCK_START,
    `if [ -z "$FLUTTER_COMPILE_SDK" ]; then`,
    `  export PATH="${sdkDir}/bin:$PATH"`,
    `  export PATH="${sdkDir}/bin/cache/dart-sdk/bin:$PATH"`,
    `  export PUB_CACHE="${pubCachePath}"`,
    `else`,
    `  export PATH="$FLUTTER_COMPILE_SDK/bin:$FLUTTER_COMPILE_SDK/bin/cache/dart-sdk/bin:$PATH"`,
    `  export PUB_CACHE="$FLUTTER_COMPILE_SDK/.pub-cache"`,
    `fi`,
    SDK_PATH_BLOCK_END,
    "",
  ].join("\n");
}

/** Write the SDK PATH block into the env file, replacing any existing block. */
function updateShellConfigPath(sdkDir: string, pubCachePath: string): void {
  if (!isFlutterSdk(sdkDir)) {
    console.warn(
      `Refusing to update shell config: "${sdkDir}" is not a valid Flutter SDK.`,
    );
    return;
  }

  ensureSourceLine();

  const envFile = envFilePath();
  let contents = "";
  try {
    contents = fs.readFileSync(envFile, "utf-8");
  } catch {
    // File doesn't exist yet — start with empty string
  }

  // Remove existing block
  contents = contents.replace(SDK_PATH_BLOCK_RE, "");

  // Append new block
  contents += buildSdkPathBlock(sdkDir, pubCachePath);

  const dir = path.dirname(envFile);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(envFile, contents, "utf-8");
}

/** Remove the SDK PATH block from the env file and legacy shell RC. */
function removeShellConfigPath(): void {
  // Clean env file
  const envFile = envFilePath();
  if (fs.existsSync(envFile)) {
    let contents = fs.readFileSync(envFile, "utf-8");
    contents = contents.replace(SDK_PATH_BLOCK_RE, "");
    fs.writeFileSync(envFile, contents, "utf-8");
  }

  // Migration: also strip legacy block from shell RC
  const configFile = shellConfigPath();
  if (fs.existsSync(configFile)) {
    let contents = fs.readFileSync(configFile, "utf-8");
    if (SDK_PATH_BLOCK_RE.test(contents)) {
      contents = contents.replace(SDK_PATH_BLOCK_RE, "");
      fs.writeFileSync(configFile, contents, "utf-8");
    }
  }
}

// ---------------------------------------------------------------------------
// Env file migration
// ---------------------------------------------------------------------------

/**
 * Migrate the env file: rewrite old-style SDK blocks with the new
 * FLUTTER_COMPILE_SDK-guarded template. No-op if already migrated.
 *
 * Called once on extension activation so that existing users get the
 * project-pin fix without having to re-set their global SDK.
 */
export function migrateEnvFile(): void {
  const envFile = envFilePath();
  let contents = "";
  try {
    contents = fs.readFileSync(envFile, "utf-8");
  } catch {
    return; // No env file, nothing to migrate
  }

  // Nothing to migrate if no SDK block or already guarded
  if (
    !SDK_PATH_BLOCK_RE.test(contents) ||
    contents.includes("FLUTTER_COMPILE_SDK")
  ) {
    return;
  }

  // Resolve the global SDK to rewrite the block with the new template
  const globalVersion = readRcConfigValue(GLOBAL_SDK_KEY);
  if (!globalVersion || !globalVersion.trim()) {
    return;
  }

  // Use fallback scan to handle directories with trailing whitespace
  const trimmed = globalVersion.trim();
  let sdkDir = sdkVersionPath(trimmed);
  if (!fs.existsSync(sdkDir)) {
    // Fallback: scan versions dir for trimmed name match, renaming if needed
    const vDir = versionsDir();
    let found = false;
    if (fs.existsSync(vDir)) {
      try {
        for (const d of fs.readdirSync(vDir, { withFileTypes: true })) {
          if (d.isDirectory() && d.name.trim() === trimmed && d.name !== trimmed) {
            try {
              fs.renameSync(path.join(vDir, d.name), sdkDir);
            } catch {
              sdkDir = path.join(vDir, d.name);
            }
            found = true;
            break;
          }
        }
      } catch {
        // ignore
      }
    }
    if (!found) {
      return;
    }
  }

  const pubCache = path.join(sdkDir, ".pub-cache");
  updateShellConfigPath(sdkDir, pubCache);
}

// ---------------------------------------------------------------------------
// Git helpers
// ---------------------------------------------------------------------------

/**
 * Returns true if the path contains a usable Flutter SDK (`bin/flutter`).
 *
 * Checks for the flutter executable rather than `.git/HEAD` so that SDKs
 * installed from release archives (no `.git`) still work.
 */
function isFlutterSdk(sdkPath: string): boolean {
  const flutter =
    process.platform === "win32"
      ? path.join(sdkPath, "bin", "flutter.bat")
      : path.join(sdkPath, "bin", "flutter");
  try {
    return fs.statSync(flutter).isFile();
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// Default symlink helpers
// ---------------------------------------------------------------------------

/** Create or update the `default` symlink to point to `sdkDir`. */
function updateDefaultSdkLink(sdkDir: string): void {
  const linkPath = path.join(versionsDir(), DEFAULT_SDK_LINK);
  try {
    const stat = fs.lstatSync(linkPath);
    if (stat.isSymbolicLink()) {
      fs.unlinkSync(linkPath);
    } else if (stat.isDirectory()) {
      return; // real directory — don't touch
    }
  } catch {
    // does not exist — fine
  }
  fs.symlinkSync(sdkDir, linkPath, "dir");
}

/** Remove the `default` symlink if it exists. */
function removeDefaultSdkLink(): void {
  const linkPath = path.join(versionsDir(), DEFAULT_SDK_LINK);
  try {
    const stat = fs.lstatSync(linkPath);
    if (stat.isSymbolicLink()) {
      fs.unlinkSync(linkPath);
    }
  } catch {
    // does not exist — fine
  }
}

// ---------------------------------------------------------------------------
// Native SDK Backend
// ---------------------------------------------------------------------------

export class NativeSdkBackend implements SdkBackend {
  async listSdks(projectRoot?: string): Promise<SdkEntry[]> {
    const vDir = versionsDir();
    if (!fs.existsSync(vDir)) {
      return [];
    }

    let entries: { name: string; dirName: string }[];
    try {
      entries = fs
        .readdirSync(vDir, { withFileTypes: true })
        .filter((d) => d.isDirectory() && d.name !== DEFAULT_SDK_LINK)
        .map((d) => ({ name: d.name.trim(), dirName: d.name }))
        .sort((a, b) => a.name.localeCompare(b.name));
    } catch {
      return [];
    }

    if (entries.length === 0) {
      return [];
    }

    const globalVersion = readRcConfigValue(GLOBAL_SDK_KEY)?.trim();

    let projectVersion: string | undefined;
    const root = projectRoot ?? vscode.workspace.workspaceFolders?.[0]?.uri.fsPath;
    if (root) {
      const fvFile = path.join(root, FLUTTER_VERSION_FILE);
      if (fs.existsSync(fvFile)) {
        projectVersion = fs.readFileSync(fvFile, "utf-8").trim() || undefined;
      }
    }

    const sdks: SdkEntry[] = entries.map((e) => ({
      version: e.name,
      path: path.join(vDir, e.dirName),
      global: e.name === globalVersion,
      project: e.name === projectVersion,
      contributor: false,
    }));

    // Check for contributor (compiled) environment
    const compiledDir = path.join(homeDir(), "flutter_compile", "flutter");
    if (fs.existsSync(compiledDir)) {
      sdks.push({
        version: "compiled",
        path: compiledDir,
        global: "compiled" === globalVersion,
        project: "compiled" === projectVersion,
        contributor: true,
      });
    }

    return sdks;
  }

  async getGlobalSdkVersion(): Promise<string | undefined> {
    const value = readRcConfigValue(GLOBAL_SDK_KEY);
    return value && value.trim() ? value.trim() : undefined;
  }

  async setGlobalSdk(version: string): Promise<void> {
    const sdkDir = await this.getSdkPath(version);
    if (!sdkDir || !isFlutterSdk(sdkDir)) {
      throw new Error(`SDK "${version}" is not installed.`);
    }
    writeRcConfigKey(GLOBAL_SDK_KEY, version);

    // Update shell config so new terminal sessions use this SDK
    const pubCachePath = path.join(sdkDir, ".pub-cache");
    updateShellConfigPath(sdkDir, pubCachePath);

    // Update the `default` symlink
    updateDefaultSdkLink(sdkDir);
  }

  installSdkInTerminal(version: string): void {
    const target = sdkVersionPath(version);
    const vDir = versionsDir();
    const pubCache = path.join(target, ".pub-cache");

    // Build a shell script that clones, checks out, and caches the SDK
    const commands = [
      `mkdir -p "${vDir}"`,
      `git clone "${FLUTTER_GIT_URL}" "${target}"`,
      `cd "${target}" && git checkout ${version}`,
      `export PUB_CACHE="${pubCache}"`,
      `"${target}/bin/flutter" --version`,
    ];

    const terminal = vscode.window.createTerminal("Flutter SDK Install");
    terminal.sendText(commands.join(" && "));
    terminal.show();
  }

  async removeSdk(version: string): Promise<void> {
    if (["compiled", "engine"].includes(version.trim())) {
      throw new Error(
        '"compiled" is the contributor environment — remove it with ' +
          '"Flutter Compile: Uninstall", not the SDK manager.'
      );
    }
    const sdkDir = sdkVersionPath(version);
    if (!fs.existsSync(sdkDir)) {
      throw new Error(`SDK "${version}" is not installed.`);
    }

    fs.rmSync(sdkDir, { recursive: true, force: true });

    // If it was the global version, clear the config and shell PATH
    const globalVersion = readRcConfigValue(GLOBAL_SDK_KEY);
    if (globalVersion === version) {
      removeRcConfigKey(GLOBAL_SDK_KEY);
      removeShellConfigPath();
      removeDefaultSdkLink();
    }
  }

  async pinToProject(version: string, projectRoot: string): Promise<void> {
    const sdkPath = await this.getSdkPath(version);
    if (!sdkPath || !isFlutterSdk(sdkPath)) {
      throw new Error(`SDK "${version}" is not installed.`);
    }
    const filePath = path.join(projectRoot, FLUTTER_VERSION_FILE);
    fs.writeFileSync(filePath, `${version}\n`, "utf-8");
  }

  async unpinFromProject(projectRoot: string): Promise<void> {
    const filePath = path.join(projectRoot, FLUTTER_VERSION_FILE);
    try {
      fs.rmSync(filePath);
    } catch {
      // File doesn't exist — nothing to do
    }
  }

  async getSdkPath(version: string): Promise<string | undefined> {
    const trimmed = version.trim();
    // Contributor environment: the from-source checkout created by
    // `install flutter`, selectable under the fixed name "compiled".
    if (trimmed === "compiled") {
      const checkout = path.join(homeDir(), "flutter_compile", "flutter");
      return fs.existsSync(checkout) ? checkout : undefined;
    }
    const sdkPath = sdkVersionPath(trimmed);
    if (fs.existsSync(sdkPath)) {
      return sdkPath;
    }

    // Fallback: scan versions dir for a directory whose trimmed name matches.
    // When found, rename to the canonical name so trailing whitespace doesn't
    // leak into shell config files.
    const vDir = versionsDir();
    if (fs.existsSync(vDir)) {
      try {
        for (const d of fs.readdirSync(vDir, { withFileTypes: true })) {
          if (d.isDirectory() && d.name.trim() === trimmed && d.name !== trimmed) {
            try {
              fs.renameSync(path.join(vDir, d.name), sdkPath);
            } catch {
              return path.join(vDir, d.name); // rename failed — return raw path
            }
            return sdkPath;
          }
        }
      } catch {
        // ignore
      }
    }

    // Also check the compiled environment
    if (trimmed === "compiled") {
      const compiledDir = path.join(homeDir(), "flutter_compile", "flutter");
      if (fs.existsSync(compiledDir)) {
        return compiledDir;
      }
    }
    return undefined;
  }

  async isSdkInstalled(version: string): Promise<boolean> {
    const sdkPath = await this.getSdkPath(version);
    return sdkPath != null && isFlutterSdk(sdkPath);
  }
}

File: extensions/intellij/src/main/kotlin/com/flutterplaza/fluttercompile/sdk/NativeSdkBackend.kt
package com.flutterplaza.fluttercompile.sdk

import com.flutterplaza.fluttercompile.Constants
import com.flutterplaza.fluttercompile.cli.SdkEntry
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.CapturingProcessHandler
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

/** Native SDK backend that manages Flutter SDKs directly via filesystem + git. */
class NativeSdkBackend : SdkBackend {

    companion object {
        private val LOG = Logger.getInstance(NativeSdkBackend::class.java)

        private fun homeDir(): String = System.getProperty(Constants.SYS_USER_HOME)
        private fun versionsDir(): File = File(homeDir(), Constants.SDK_VERSIONS_REL)
        private fun sdkVersionDir(version: String): File = File(versionsDir(), version)

        /** Resolve the actual SDK directory — compiled SDK lives at ~/flutter_compile/flutter. */
        private fun resolveSdkDir(version: String): File {
            if (version == Constants.COMPILED_VERSION) return File(homeDir(), Constants.COMPILED_SDK_REL)
            return sdkVersionDir(version)
        }

        /** Create or update the `default` symlink to point to [sdkDir]. */
        private fun updateDefaultSdkLink(sdkDir: File) {
            val linkPath = Paths.get(versionsDir().absolutePath, Constants.DEFAULT_SDK_LINK)
            if (Files.exists(linkPath, java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
                if (Files.isSymbolicLink(linkPath)) {
                    Files.delete(linkPath)
                } else if (Files.isDirectory(linkPath, java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
                    return // real directory — don't touch
                }
            }
            try { Files.createSymbolicLink(linkPath, sdkDir.toPath()) } catch (e: Exception) {
                LOG.warn("Failed to create default SDK symlink", e)
            }
        }

        /** Remove the `default` symlink if it exists. */
        private fun removeDefaultSdkLink() {
            val linkPath = Paths.get(versionsDir().absolutePath, Constants.DEFAULT_SDK_LINK)
            if (Files.isSymbolicLink(linkPath)) {
                try { Files.delete(linkPath) } catch (e: Exception) {
                    LOG.warn("Failed to remove default SDK symlink", e)
                }
            }
        }
    }

    override fun listSdks(projectPath: String?): List<SdkEntry> {
        val vDir = versionsDir()
        if (!vDir.exists()) return emptyList()

        val entries = vDir.listFiles()
            ?.filter { it.isDirectory && it.name != Constants.DEFAULT_SDK_LINK }
            ?.sortedBy { it.name }
            ?: return emptyList()

        if (entries.isEmpty()) return emptyList()

        val globalVersion = RcConfig.readValue(Constants.GLOBAL_SDK_KEY)?.trim()

        var projectVersion: String? = null
        if (projectPath != null) {
            val fvFile = File(projectPath, Constants.FLUTTER_VERSION_FILE)
            if (fvFile.exists()) {
                projectVersion = fvFile.readText().trim().ifEmpty { null }
            }
        }

        val sdks = entries.map { dir ->
            val trimmedName = dir.name.trim()
            SdkEntry(
                version = trimmedName,
                path = dir.absolutePath,
                global = trimmedName == globalVersion,
                project = trimmedName == projectVersion,
                contributor = false,
            )
        }.toMutableList()

        // Check for contributor (compiled) environment
        val compiledDir = File(homeDir(), Constants.COMPILED_SDK_REL)
        if (compiledDir.exists()) {
            sdks.add(
                SdkEntry(
                    version = Constants.COMPILED_VERSION,
                    path = compiledDir.absolutePath,
                    global = Constants.COMPILED_VERSION == globalVersion,
                    project = Constants.COMPILED_VERSION == projectVersion,
                    contributor = true,
                )
            )
        }

        return sdks
    }

    override fun getGlobalSdkVersion(): String? {
        val value = RcConfig.readValue(Constants.GLOBAL_SDK_KEY)?.trim()
        return if (value.isNullOrEmpty()) null else value
    }

    override fun setGlobalSdk(version: String): Boolean {
        val sdkPath = getSdkPath(version) ?: run {
            LOG.warn("Cannot set global: SDK '$version' is not installed")
            return false
        }
        val sdkDir = File(sdkPath)
        if (!isFlutterSdk(sdkDir)) {
            LOG.warn("Cannot set global: SDK '$version' is not a valid Flutter SDK")
            return false
        }
        RcConfig.writeValue(Constants.GLOBAL_SDK_KEY, version.trim())

        // Update shell config so new terminal sessions use this SDK
        val pubCachePath = File(sdkDir, ".pub-cache").absolutePath
        ShellConfig.updatePath(sdkDir.absolutePath, pubCachePath)

        // Update the `default` symlink
        updateDefaultSdkLink(sdkDir)

        return true
    }

    override fun installSdk(version: String, indicator: ProgressIndicator?): Boolean {
        val target = sdkVersionDir(version)

        if (isValidGitRepo(target)) {
            LOG.info("SDK '$version' already installed at ${target.absolutePath}")
            return true
        }

        return try {
            // Ensure versions directory exists
            versionsDir().mkdirs()

            indicator?.text = "Cloning Flutter SDK '$version'..."
            runGit("clone", Constants.FLUTTER_GIT_URL, target.absolutePath)

            indicator?.text = "Checking out '$version'..."
            runGitIn(target, "checkout", version)

            indicator?.text = "Caching Flutter SDK artifacts..."
            val flutterBin = File(target, "bin/flutter").absolutePath
            val pubCache = File(target, ".pub-cache").absolutePath
            val cmd = GeneralCommandLine(flutterBin, "--version")
                .withWorkDirectory(target)
                .withEnvironment("PUB_CACHE", pubCache)
                .withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.CONSOLE)
            val handler = CapturingProcessHandler(cmd)
            handler.runProcess(300_000)

            // Auto-set global if no global version exists
            if (getGlobalSdkVersion() == null) {
                RcConfig.writeValue(Constants.GLOBAL_SDK_KEY, version)
                ShellConfig.updatePath(target.absolutePath, pubCache)
                updateDefaultSdkLink(target)
            }

            true
        } catch (e: Exception) {
            LOG.warn("Failed to install SDK '$version'", e)
            // Clean up partial install
            if (target.exists()) {
                target.deleteRecursively()
            }
            false
        }
    }

    override fun removeSdk(version: String): Boolean {
        val sdkDir = sdkVersionDir(version)
        if (!sdkDir.exists()) {
            LOG.warn("SDK '$version' not installed")
            return false
        }

        return try {
            sdkDir.deleteRecursively()

            // Clear global config and shell PATH if this was the global version
            val globalVersion = RcConfig.readValue(Constants.GLOBAL_SDK_KEY)?.trim()
            if (globalVersion == version) {
                RcConfig.removeKey(Constants.GLOBAL_SDK_KEY)
                ShellConfig.removePath()
                removeDefaultSdkLink()
            }

            true
        } catch (e: Exception) {
            LOG.warn("Failed to remove SDK '$version'", e)
            false
        }
    }

    override fun pinToProject(version: String, projectPath: String): Boolean {
        val sdkPath = getSdkPath(version) ?: run {
            LOG.warn("Cannot pin: SDK '$version' is not installed")
            return false
        }
        if (!isFlutterSdk(File(sdkPath))) {
            LOG.warn("Cannot pin: SDK '$version' is not a valid Flutter SDK")
            return false
        }
        return try {
            File(projectPath, Constants.FLUTTER_VERSION_FILE).writeText("${version.trim()}\n")
            true
        } catch (e: Exception) {
            LOG.warn("Failed to pin SDK '$version' to project", e)
            false
        }
    }

    override fun unpinFromProject(projectPath: String): Boolean {
        val fvFile = File(projectPath, Constants.FLUTTER_VERSION_FILE)
        if (!fvFile.exists()) return true // already unpinned
        return try {
            fvFile.delete()
        } catch (e: Exception) {
            LOG.warn("Failed to unpin SDK from project", e)
            false
        }
    }

    override fun getSdkPath(version: String): String? {
        val trimmed = version.trim()
        val sdkDir = sdkVersionDir(trimmed)
        if (sdkDir.exists()) return sdkDir.absolutePath

        // Fallback: scan versions dir for a directory whose trimmed name matches.
        // When found, rename to the canonical name so trailing whitespace doesn't
        // leak into shell config files.
        val vDir = versionsDir()
        if (vDir.exists()) {
            vDir.listFiles()?.forEach { dir ->
                if (dir.isDirectory && dir.name.trim() == trimmed && dir.name != trimmed) {
                    return try {
                        dir.renameTo(sdkDir)
                        sdkDir.absolutePath
                    } catch (_: Exception) {
                        dir.absolutePath // rename failed — return raw path
                    }
                }
            }
        }

        if (trimmed == Constants.COMPILED_VERSION) {
            val compiledDir = File(homeDir(), Constants.COMPILED_SDK_REL)
            if (compiledDir.exists()) return compiledDir.absolutePath
        }

        return null
    }

    override fun isSdkInstalled(version: String): Boolean {
        val sdkPath = getSdkPath(version) ?: return false
        return isFlutterSdk(File(sdkPath))
    }

    private fun isValidGitRepo(dir: File): Boolean {
        return dir.exists() && File(dir, Constants.GIT_HEAD_FILE).exists()
    }

    /**
     * Returns true if [dir] contains a usable Flutter SDK (`bin/flutter`).
     *
     * Less strict than [isValidGitRepo] — works for SDKs installed from
     * release archives or where `.git` was cleaned up.
     */
    private fun isFlutterSdk(dir: File): Boolean {
        val osName = System.getProperty(Constants.SYS_OS_NAME, "").lowercase()
        val flutter = if (osName.contains(Constants.OS_WINDOWS_MARKER)) {
            File(dir, "bin${File.separator}flutter.bat")
        } else {
            File(dir, "bin${File.separator}flutter")
        }
        return flutter.exists()
    }

    private fun runGit(vararg args: String) {
        val cmd = GeneralCommandLine(Constants.CMD_GIT, *args)
            .withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.CONSOLE)
        val handler = CapturingProcessHandler(cmd)
        val result = handler.runProcess(300_000)
        if (result.exitCode != 0) {
            throw RuntimeException("git ${args.joinToString(" ")} failed: ${result.stderr}")
        }
    }

    private fun runGitIn(dir: File, vararg args: String) {
        val cmd = GeneralCommandLine(Constants.CMD_GIT, *args)
            .withWorkDirectory(dir)
            .withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.CONSOLE)
        val handler = CapturingProcessHandler(cmd)
        val result = handler.runProcess(300_000)
        if (result.exitCode != 0) {
            throw RuntimeException("git ${args.joinToString(" ")} failed: ${result.stderr}")
        }
    }
}
